git/libgit2/transport/http.go

```diff
@@ -0,0 +1,479 @@
+
+	if t.httpTransport != nil {
+		t.logger.V(logger.TraceLevel).Info("release http transport back to pool")
+
+		transport.Release(t.httpTransport)
+		t.httpTransport = nil
+	}
+}
+
+type httpSmartSubtransportStream struct {
+	owner       *httpSmartSubtransport
+	client      *http.Client
+	req         *http.Request
+	resp        *http.Response
+	reader      *io.PipeReader
+	writer      *io.PipeWriter
+	sentRequest bool
+	recvReply   sync.WaitGroup
+	httpError   error
+	m           sync.RWMutex
+}
+
+var _ git2go.SmartSubtransportStream = &httpSmartSubtransportStream{}
+
+func newManagedHttpStream(owner *httpSmartSubtransport, req *http.Request, client *http.Client) *httpSmartSubtransportStream {
+	r, w := io.Pipe()
+	return &httpSmartSubtransportStream{
+		owner:  owner,
+		client: client,
+		req:    req,
+		reader: r,
+		writer: w,
+	}
+}
+
+func (self *httpSmartSubtransportStream) Read(buf []byte) (int, error) {
+	if err := self.writer.Close(); err != nil {
+		return 0, err
+	}
+
+	self.recvReply.Wait()
+
+	self.m.RLock()
+	err := self.httpError
+	self.m.RUnlock()
+
+	if err != nil {
+		return 0, self.httpError
+	}
+	return self.resp.Body.Read(buf)
+}
+
+func (self *httpSmartSubtransportStream) Write(buf []byte) (int, error) {
+	self.m.RLock()
+	err := self.httpError
+	self.m.RUnlock()
+
+	if err != nil {
+		return 0, self.httpError
+	}
+	return self.writer.Write(buf)
+}
+
+func (self *httpSmartSubtransportStream) Free() {
+	if self.resp != nil {
+		self.owner.logger.V(logger.TraceLevel).Info("httpSmartSubtransportStream.Free()")
+
+		if self.resp.Body != nil {
+			// ensure body is fully processed and closed
+			// for increased likelihood of transport reuse in HTTP/1.x.
+			// it should not be a problem to do this more than once.
+			if _, err := io.Copy(io.Discard, self.resp.Body); err != nil {
+				self.owner.logger.V(logger.TraceLevel).Error(err, "cannot discard response body")
+			}
+
+			if err := self.resp.Body.Close(); err != nil {
+				self.owner.logger.V(logger.TraceLevel).Error(err, "cannot close response body")
+			}
+		}
+	}
+}
+
+func (self *httpSmartSubtransportStream) sendRequestBackground() {
+	go func() {
+		err := self.sendRequest()
+
+		self.m.Lock()
+		self.httpError = err
+		self.m.Unlock()
+	}()
+}
+
+func (self *httpSmartSubtransportStream) sendRequest() error {
+	defer self.recvReply.Done()
+	self.resp = nil
+
+	var resp *http.Response
+	var err error
+	var content []byte
+
+	for {
+		req := &http.Request{
+			Method: self.req.Method,
+			URL:    self.req.URL,
+			Header: self.req.Header,
+		}
+		req = req.WithContext(self.owner.ctx)
+
+		if req.Method == "POST" {
+			if len(content) == 0 {
+				// a copy of the request body needs to be saved so
+				// it can be reused in case of redirects.
+				if content, err = io.ReadAll(self.reader); err != nil {
+					return err
+				}
+			}
+			req.Body = io.NopCloser(bytes.NewReader(content))
+			req.ContentLength = int64(len(content))
```

Comment:
Compared to the previous implementation, this is different. As far as I know and remember, this piece of code was taken from the upstream implementation of this in git2go.
Based on net/http docs:
> // If body is of type *bytes.Buffer, *bytes.Reader, or
> // *strings.Reader, the returned request's ContentLength is set to its
> // exact value (instead of -1), GetBody is populated (so 307 and 308
> // redirects can replay the body), and Body is set to NoBody if the
> // ContentLength is 0.

and some web search, using `NopCloser()` results in content length to not be set. And not setting content length can be an issue with HTTP/2. Was this causing an issue anywhere?

I'm finding it hard to find reasons for certain changes in the implementation when compared with what's in SC and their implications.
Don't think we need to leave in-line comment about everything, but not sure what's a better way to convey the reasons for the changes now and for the future.